src/entities/addLiquidity/addLiquidityV3/index.ts

```diff
@@ -0,0 +1,154 @@
+                    input,
+                    poolState.address,
+                    input.bptOut.rawAmount,
+                );
+                amountsIn = sortedTokens.map((t, i) =>
+                    TokenAmount.fromRawAmount(t, maxAmountsIn[i]),
+                );
+                tokenInIndex = sortedTokens.findIndex((t) =>
+                    t.isSameAddress(input.tokenIn),
+                );
+                break;
+            }
+        }
+
+        const output: AddLiquidityBaseQueryOutput = {
+            poolType: poolState.type,
+            poolId: poolState.id,
+            addLiquidityKind: input.kind,
+            bptOut,
+            amountsIn,
+            fromInternalBalance: input.fromInternalBalance ?? false,
+            balancerVersion: 3,
+            tokenInIndex,
+        };
+
+        return output;
+    }
+
+    buildCall(input: AddLiquidityBaseCall): AddLiquidityBuildOutput {
+        const amounts = getAmountsCall(input);
+        let call: Hex;
+        switch (input.addLiquidityKind) {
+            case AddLiquidityKind.Proportional:
+                throw addLiquidityProportionalUnavailableError;
+            case AddLiquidityKind.Unbalanced:
+                {
+                    call = encodeFunctionData({
+                        abi: balancerRouterAbi,
+                        functionName: 'addLiquidityUnbalanced',
+                        args: [
+                            input.poolId,
+                            input.amountsIn.map((a) => a.amount),
+                            amounts.minimumBpt,
+                            input.wethIsEth,
+                            '0x',
+                        ],
+                    });
+                }
+                break;
+            case AddLiquidityKind.SingleToken:
+                {
+                    // just a sanity check as this is already checked in InputValidator
+                    if (input.tokenInIndex === undefined) {
+                        throw addLiquiditySingleTokenShouldHaveTokenInIndexError;
+                    }
+                    call = encodeFunctionData({
+                        abi: balancerRouterAbi,
+                        functionName: 'addLiquiditySingleTokenExactOut',
+                        args: [
+                            input.poolId,
+                            input.amountsIn[input.tokenInIndex].token.address,
+                            input.amountsIn[input.tokenInIndex].amount,
+                            input.bptOut.amount,
+                            input.wethIsEth,
+                            '0x',
+                        ],
+                    });
+                }
+                break;
+        }
+
+        let value = 0n;
+        if (input.wethIsEth) {
+            const wethInput = input.amountsIn.find(
+                (a) => a.token.address === NATIVE_ASSETS[input.chainId].wrapped,
+            );
+            value = wethInput?.amount ?? 0n;
```

Comment:
Do you think we should throw if wethInput doesn't exist?

Reply:
Yeap. It would be odd if the flag `wethIsEth` is set to true and `wethInput` doesn't exist. I agree it's safer to throw and let the user know there's something wrong. 👍

Reply:
done ✅